fastestimator/trace/io/model_saver.py

```diff
@@ -12,10 +12,84 @@
 # See the License for the specific language governing permissions and
 # limitations under the License.
 # ==============================================================================
+import os
+from typing import Union
 
-from fastestimator.trace.trace import Trace
+import numpy as np
+import tensorflow as tf
+import torch
+
+from fastestimator.trace import Trace
+from fastestimator.util import Data
 
 
 class ModelSaver(Trace):
-    def __init__(self, model):
-        pass
+    """Save trained model in hdf5/pt format.
+
+    Args:
+        model: Model instance defined by user (tf.keras.Model or torch.nn.Module)
+        model_name (str): Name of FE model.
+        save_dir (str): Directory to save the trained models.
+        save_freq (int, optional): Number of epochs to save models. Cannot be used with `save_best_only=True`. Defaults
+            to 1.
+        save_best (bool, str, optional): Best model saving monitor name. If True, the model loss is used. Defaults to
+            False.
+        save_best_mode (str, optional): Can be `'min'`, `'max'`, or `'auto'`. Defaults to 'max'.
+
+    """
+    def __init__(self, model: Union[tf.keras.Model, torch.nn.Module], model_name: str,
+                 save_dir: str, save_freq: int = 1, save_best: Union[bool, str]=False, save_best_mode:str='max'):
+
+        if save_best is True:
+            save_best = 'accuracy'
+
+        if isinstance(save_best, str):
+            super().__init__(inputs=save_best, mode='eval')
+        else:
+            super().__init__(mode='train')
+
+        if isinstance(model, tf.keras.Model):
+            self.ext = '.h5'
+        else:
+            self.ext = '.pt'
+
+        self.model = model
+        self.model_name=model_name
+        self.save_dir = save_dir
+        self.save_best = save_best
+        self.save_best_mode = save_best_mode
+        self.save_freq = save_freq
+        assert isinstance(self.save_freq, int), "save_freq must be integer"
+        if self.save_best_mode == "min":
+            self.best = np.Inf
+            self.monitor_op = np.less
+        elif self.save_best_mode == "max":
+            self.best = -np.Inf
+            self.monitor_op = np.greater
+        else:
+            raise ValueError("save_best_mode must be either 'min' or 'max'")
+        self.model = model
+
+    def on_begin(self, data: Data):
+        if self.save_dir:
```

Comment:
this if check should be an assert in the init instead. no point in a model saver if there's no place to save the model